testenv/utils/ssh.go

```diff
@@ -0,0 +1,299 @@
+	}
+
+	return nil
+}
+
+// Load is responsible for loading the ssh keypair from filesystem into SSHKeyPair.
+func (p *SSHKeyPair) Load(privKeyFile, pubKeyFile string) error {
+	var err error
+
+	// Read the private key file.
+	p.PrivateKey, err = os.ReadFile(privKeyFile)
+	if err != nil {
+		return fmt.Errorf("failed to read private key: %w", err)
+	}
+
+	// Read the public key file.
+	p.PublicKey, err = os.ReadFile(pubKeyFile)
+	if err != nil {
+		return fmt.Errorf("failed to read public key: %w", err)
+	}
+
+	return nil
+}
+
+// GenerateSSHKeyPair generates a new SSH key pair.
+func GenerateSSHKeyPair() (*SSHKeyPair, error) {
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048) //nolint:gomnd
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate private key: %w", err)
+	}
+
+	privateKeyPEM := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	}
+	var privateKeyData []byte
+	b := bytes.NewBuffer(privateKeyData)
+
+	if err = pem.Encode(b, privateKeyPEM); err != nil {
+		return nil, fmt.Errorf("failed to encode private key: %w", err)
+	}
+
+	publicKey, err := ssh.NewPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create public key: %w", err)
+	}
+
+	return &SSHKeyPair{
+		PublicKey:  ssh.MarshalAuthorizedKey(publicKey),
+		PrivateKey: b.Bytes(),
+	}, nil
+}
+
+type SSHForwardCallback func(ctx context.Context, err error) error
+
+type SSHForwardInput struct {
+	PrivateKey    []byte
+	User          string
+	Host          string
+	Port          int
+	LocalPort     int
+	RemoteAddress string
+	RemotePort    int
+
+	Callback SSHForwardCallback
+}
+
+func (i *SSHForwardInput) HostAddressPort() string {
+	return fmt.Sprintf("%s:%d", i.Host, i.Port)
+}
+
+func (i *SSHForwardInput) LocalAddressPort() string {
+	return fmt.Sprintf("localhost:%d", i.LocalPort)
+}
+
+func (i *SSHForwardInput) RemoteAddressPort() string {
+	if i.RemoteAddress == "" {
+		i.RemoteAddress = "localhost"
+	}
+
+	return fmt.Sprintf("%s:%d", i.RemoteAddress, i.RemotePort)
+}
+
+type SSHPortForward struct {
+	input        *SSHForwardInput
+	clientConfig ssh.ClientConfig
+
+	cancel context.CancelFunc
+	stop   atomic.Bool
+}
+
+func (f *SSHPortForward) Start(ctx context.Context) error {
+	ctx, f.cancel = context.WithCancel(ctx)
+
+	// Dial the remote server.
+	client, err := ssh.Dial("tcp", f.input.HostAddressPort(), &f.clientConfig)
+	if err != nil {
+		return fmt.Errorf("failed to dial:%w", err)
+	}
+
+	// Listen on local port.
+	listener, err := net.Listen("tcp", f.input.LocalAddressPort())
+	if err != nil {
+		return fmt.Errorf("failed to listen: %w", err)
+	}
+
+	// TODO: wait until the BannerCallBack is invoked
+
+	go func() {
+		var callbackErr error
+		defer client.Close()
+		defer listener.Close()
+
+		for callbackErr == nil && !f.stop.Load() {
+			func() {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
```

Comment:
this would need to be outside of the function (inside the loop), otherwise we might have just the loop running without returning. 
```go
for condition {
	select {} // allows it to return if the ctx is done
	func() { } // executes the logic to stream between ssh and host
}

```

Reply:
fixed 👍 thank you so much for reviewing this!